src/assert.rs

```diff
@@ -250,13 +274,17 @@ impl Assert {
     /// ```
     pub fn execute(self) -> Result<()> {
         let cmd = &self.cmd[0];
+
         let args: Vec<_> = self.cmd.iter().skip(1).collect();
         let mut command = Command::new(cmd);
         let command = command
             .stdin(Stdio::piped())
             .stdout(Stdio::piped())
-            .stderr(Stdio::piped());
-        let command = command.args(&args);
+            .stderr(Stdio::piped())
+            .env_clear()
```

Comment:
There a reason we unconditionally clear and then re-add the `vars` we cleared via `compile`?

Reply:
In fact we need the `env_clear()` because [`.envs`](https://doc.rust-lang.org/src/std/process.rs.html#468) doesn't clear it or replace the whole variables.

Reply:
I think I'm missing something.  I had assumed we'd only call `env_clear()` when `inherit == false`.  Its unclear to me why we need `env_clear` when `inherit == true`.

Reply:
By clearing the environment before adding our own environment, we don't have to check if inherit is true nor false in execute command. We just want to clear and add the appropriate environment. That's the environment role to know when to add the existing env or not.

If inherit is set to true, the `compile` method will return a merge of actual environment (`env::vars()`) and our customized variables. I think it's nicer to have a compile method returning what our environment will look like. But feel free to propose something else I'm ok with it :)

Reply:
tl;dr move along, these aren't the droids you're looking for.

>  I think it's nicer to have a compile method returning what our environment will look like. 

That is nice.  It helps with testing, etc.

I just assumed that `Assert` would directly access the members of `Environment` and there wouldn't be a need for a public `.compile()`.

Recreating an inherited environment is more work but I'm assuming it doesn't matter in this case, despite it being against "you pay for what you use" (adds cost to people not using these new features).

Reply:
I do like the separation of concerns. `execute` shouldn't need to care about what the Environment contains or how its implemented. And while we could make `inherit` be `pub(crate)` (so we don't even need a getter), I do think this way is more elegant.

But: I'm a bit wary of this. "Recreating an inherited environment" my sound better than it is—IIRC the `std::env::vars()` step includes decoding env vars as UTF-8, which is lossy. Is that the case? Can we add a test for that?

Reply:
@killercup how do you see this test?

Reply:
Set an env var to something that is not UTF-8. How would I get a byte slice like that? `head -c 42 /dev/urandom | rcat --quote` (with [this](https://github.com/frewsxcv/rcat) rcat) gives me:

```rust
b"\xf4\xa9\xd5\x8d\xc0\xe8\x00\xff\x1dqz\x86k\x83\xcb~\xc1\x18!\xfa\x01J\x1c\\:\x18)%\xd4\xa6^7\xb4\x93\x80.\x9a\x9c\xe1\xcf\xe2,"
```

which is [not valid UTF-8](https://play.rust-lang.org/?gist=d94adb121157d1d9d6e7479ca650882c&version=stable).

Reply:
Yes, but `vars()` use UTF-8 variables, do we really want to support not valid UTF-8? I don't think..

Reply:
That's exactly what I'm saying: I'm fine with only allowing our users to add UTF-8 vars but we should not destroy valid, inherited env vars. So, calling `vars` at all is probably not a good call, and `vars_os` is the better option. Btw, you can find the impl of how `std::process::Command` deals with env vars on Unix [here](https://github.com/rust-lang/rust/blob/0217315bf23edee385a7042b5a23b3e7e376820c/src/libstd/sys/unix/process/process_common.rs#L124) (it uses a `HashMap<OsString, (usize, CString)>` to store env vars).

How much work would it be to allow `OsStr` instead of `String` in our APIs?

Reply:
Can you still add a test with the non-UTF-8 env var value?